yarnitor/background/worker.py

```diff
@@ -34,260 +38,390 @@
 THREADPOOL_TIMEOUT = 120
 # Sentinel state used when we fail to query the application for its state
 NON_RESPONSIVE_STATE = 'NON_RESPONSIVE'
+# Maximum number of YARN high-availability (HA) redirects to follow before
+# giving up.
+MAX_HA_REDIRECTS = 5
 
+# Global threadpool for running async tasks
+threadpool = concurrent.futures.ThreadPoolExecutor(THREADPOOL_SIZE)
+# Shut down the pool when the process is exiting
+atexit.register(lambda: threadpool.shutdown(False))
 
 class Progress(object):
-    """Utility class for storing progress information
-    """
-
+    """Utility class for storing mutable progress information."""
     def __init__(self, name, completed=0, failed=0, running=0, total=0):
         self.name = name
         self.completed = completed
         self.failed = failed
         self.running = running
         self.total = total
 
+    def to_dict(self):
```

Comment:
Given `p = Progress(...)`

is `p.to_dict()` equivalent to `dict(p)`?

Reply:
It's an explicit version of getting a dictionary view of the object vs what was used before `p.__dict__`, which might pull in new fields introduced in the future unintentionally. Now class encapsulates what the dict representation of the object is.